### ISSUE TITLE
intermediate coalescing modes / handling exotic blend sources/modes

### ISSUE THREAD

**Issue by jonsneyers** [libjxl/libjxl#1886]
Not related to this, but another coalescing mode that would be good to have for animation display, is one where the returned frames are blended but only the crop box containing modified pixels gets returned (so it only returns kReplace frames like the default coalescing, but not necessarily canvas-sized ones). This has obvious advantages in terms of buffer sizes and would allow e.g. browsers to have a higher likelihood of being able to keep the entire animation in memory.

So I think we basically need one boolean setting to indicate whether zero-duration frames should be blended or not, and another boolean setting to indicate whether cropped frames should be returned or canvas-sized ones. We currently only have an option where it is (true,false)  (the default coalescing==true) and another where it is (false,true)  (this is what coalescing==false does).

And then we need to find a way for dealing with exotic blend source choices; probably this is something most applications will not want to deal with so it's probably best if by default anything that is not a simple "stack of layers" gets converted in something that is (we can always do the blending and turn anything into a kReplace frame that overwrites the whole canvas), and only when some new option is set, you can get the original thing which may be quite exotic.
Also it would be good to have a way to tell the decoder not to return 'exotic' blend modes (i.e. not kReplace and not kBlend), since in many cases (like Gimp) the application will already know about replace and blend but not about the other blend modes.

_Originally posted by @jonsneyers in https://github.com/libjxl/libjxl/issues/1863#issuecomment-1310085823_

**Comment by ctrlcctrlv** [comment]
As mentioned at https://github.com/libjxl/libjxl/pull/1863#issuecomment-1310757987, GIMP has many blending modes.

But @jonsneyers here's the real issue. Here are all the GIMP blend modes:

1. `GIMP_LAYER_MODE_NORMAL_LEGACY`
1. `GIMP_LAYER_MODE_DISSOLVE`
1. `GIMP_LAYER_MODE_BEHIND_LEGACY`
1. `GIMP_LAYER_MODE_MULTIPLY_LEGACY`
1. `GIMP_LAYER_MODE_SCREEN_LEGACY`
1. `GIMP_LAYER_MODE_OVERLAY_LEGACY`
1. `GIMP_LAYER_MODE_DIFFERENCE_LEGACY`
1. `GIMP_LAYER_MODE_ADDITION_LEGACY`
1. `GIMP_LAYER_MODE_SUBTRACT_LEGACY`
1. `GIMP_LAYER_MODE_DARKEN_ONLY_LEGACY`
1. `GIMP_LAYER_MODE_LIGHTEN_ONLY_LEGACY`
1. `GIMP_LAYER_MODE_HSV_HUE_LEGACY`
1. `GIMP_LAYER_MODE_HSV_SATURATION_LEGACY`
1. `GIMP_LAYER_MODE_HSL_COLOR_LEGACY`
1. `GIMP_LAYER_MODE_HSV_VALUE_LEGACY`
1. `GIMP_LAYER_MODE_DIVIDE_LEGACY`
1. `GIMP_LAYER_MODE_DODGE_LEGACY`
1. `GIMP_LAYER_MODE_BURN_LEGACY`
1. `GIMP_LAYER_MODE_HARDLIGHT_LEGACY`
1. `GIMP_LAYER_MODE_SOFTLIGHT_LEGACY`
1. `GIMP_LAYER_MODE_GRAIN_EXTRACT_LEGACY`
1. `GIMP_LAYER_MODE_GRAIN_MERGE_LEGACY`
1. `GIMP_LAYER_MODE_COLOR_ERASE_LEGACY`
1. `GIMP_LAYER_MODE_OVERLAY`
1. `GIMP_LAYER_MODE_LCH_HUE`
1. `GIMP_LAYER_MODE_LCH_CHROMA`
1. `GIMP_LAYER_MODE_LCH_COLOR`
1. `GIMP_LAYER_MODE_LCH_LIGHTNESS`
1. `GIMP_LAYER_MODE_NORMAL`
1. `GIMP_LAYER_MODE_BEHIND`
1. `GIMP_LAYER_MODE_MULTIPLY`
1. `GIMP_LAYER_MODE_SCREEN`
1. `GIMP_LAYER_MODE_DIFFERENCE`
1. `GIMP_LAYER_MODE_ADDITION`
1. `GIMP_LAYER_MODE_SUBTRACT`
1. `GIMP_LAYER_MODE_DARKEN_ONLY`
1. `GIMP_LAYER_MODE_LIGHTEN_ONLY`
1. `GIMP_LAYER_MODE_HSV_HUE`
1. `GIMP_LAYER_MODE_HSV_SATURATION`
1. `GIMP_LAYER_MODE_HSL_COLOR`
1. `GIMP_LAYER_MODE_HSV_VALUE`
1. `GIMP_LAYER_MODE_DIVIDE`
1. `GIMP_LAYER_MODE_DODGE`
1. `GIMP_LAYER_MODE_BURN`
1. `GIMP_LAYER_MODE_HARDLIGHT`
1. `GIMP_LAYER_MODE_SOFTLIGHT`
1. `GIMP_LAYER_MODE_GRAIN_EXTRACT`
1. `GIMP_LAYER_MODE_GRAIN_MERGE`
1. `GIMP_LAYER_MODE_VIVID_LIGHT`
1. `GIMP_LAYER_MODE_PIN_LIGHT`
1. `GIMP_LAYER_MODE_LINEAR_LIGHT`
1. `GIMP_LAYER_MODE_HARD_MIX`
1. `GIMP_LAYER_MODE_EXCLUSION`
1. `GIMP_LAYER_MODE_LINEAR_BURN`
1. `GIMP_LAYER_MODE_LUMA_DARKEN_ONLY`
1. `GIMP_LAYER_MODE_LUMA_LIGHTEN_ONLY`
1. `GIMP_LAYER_MODE_LUMINANCE`
1. `GIMP_LAYER_MODE_COLOR_ERASE`
1. `GIMP_LAYER_MODE_ERASE`
1. `GIMP_LAYER_MODE_MERGE`
1. `GIMP_LAYER_MODE_SPLIT`
1. `GIMP_LAYER_MODE_PASS_THROUGH`

Source: `enum GimpLayerMode`, file `/usr/include/gimp-2.0/libgimp/gimpenums.h`.

Meanwhile `libjxl` defines:
```cpp
typedef enum {
  JXL_BLEND_REPLACE = 0,
  JXL_BLEND_ADD = 1,
  JXL_BLEND_BLEND = 2,
  JXL_BLEND_MULADD = 3, 
  JXL_BLEND_MUL = 4,    
} JxlBlendMode;
```

62 vs. 5. Mapping these appropriately is hardly simple and will be highly bug prone especially when different color schemes and ICC's come into play.

**Comment by jonsneyers** [comment]
Yes, GIMP has way more blend modes than JXL, the only intersection there is that both have normal blending (and replace, though that one is done by setting the layer name in GIMP).

So we cannot map arbitrary GIMP layers to JXL. The other way around can also not be done in the general case, though we can always produce a replace layer instead.

The main issue is that JXL has much richer "dispose modes" than what GIF/APNG/WebP has and GIMP too via the layer naming convention. JXL layers can blend to arbitrary previous canvas states, not just to the current state ("dispose none" in APNG terms) and the previous state ("dispose previous" in APNG terms). So there too, we would have to convert exotic blend sources to replace frames to make it work. At least unless GIMP would be adding richer syntax to the layer name conventions so it can represent what JXL bitstreams can do.

**Comment by ctrlcctrlv** [comment]
I'm not necessarily bound to follow the layer name convention, which is all _every other GIMP plugin_ takes into account when saving: replace or combine. I could choose to take into account layer blend modes as well, if the blend mode matched a map from GIMP→JXL modes, but it'd only work when saving a JXL animation. I think it's very confusing for users.

**Comment by jonsneyers** [comment]
Yes, I think it is better to map jxl to the current gimp conventions. You could do that by using the libjxl api as it is now and doing blending yourself as needed to get rid of the exotic stuff, but it's probably better if we have a libjxl option to do that for you and produce only frames that can be directly mapped to gif/apng/webp frames. This is likely functionality that is also useful for other applications that already support gif/apng, so they can just add jxl without needing new blending primitives (but also without requiring everything to be coalesced to full canvas-sized replace-only frames).

**Comment by ctrlcctrlv** [comment]
We agree then :-)
